File: packages/sli-ui-toolkit/src/sli_ui_toolkit/ui/widgets/helpers/underline_painter.py
from dataclasses import dataclass
from typing import List, Optional, Union

from PyQt6.QtCore import QPointF, QRectF, Qt
from PyQt6.QtGui import QColor, QPen
from PyQt6.QtWidgets import QLineEdit

from sli_ui_toolkit.theme import ThemeManager

@dataclass
class UnderlineConfig:
    thickness: float = 0.15
    vertical_offset: float = 0.75
    arc_radius: float = 1.33
    alpha: Optional[int] = None
    color: Union[QColor, List[QColor], None] = None

def _widget_scale(rect) -> float:
    """Scale factor based on widget height (baseline: 32px button)."""
    h = float(rect.height())
    return max(1.0, h / 32.0)

def draw_bottom_underline(
    painter, rect, theme_manager: ThemeManager, config: UnderlineConfig | None = None
):
    cfg = config or UnderlineConfig()
    widget = painter.device()

    if theme_manager.is_dark():
        if not (widget and isinstance(widget, QLineEdit)):
            return

    if widget and hasattr(widget, "property"):
        btn_class = str(widget.property("class") or "")
        prefix = "button.primary" if btn_class == "primary" else "button.default"
    else:
        prefix = "button.default"

    if isinstance(cfg.color, list) and cfg.color:
        colors = cfg.color
    elif isinstance(cfg.color, QColor):
        colors = [cfg.color]
    else:
        colors = [QColor(theme_manager.get_color(f"{prefix}.bottom.edge"))]

    final_colors = []
    for color in colors:
        new_color = QColor(color)
        if cfg.alpha is not None:
            new_color.setAlpha(int(cfg.alpha))
        final_colors.append(new_color)

    count = len(final_colors)
    if count == 0:
        return

    scale = _widget_scale(rect)
    arc_radius = float(cfg.arc_radius) * scale
    thickness = cfg.thickness * scale
    vertical_offset = cfg.vertical_offset * scale
    base_y = float(rect.bottom()) - vertical_offset
    start_x = float(rect.left())
    end_x = float(rect.right())
    total_width = end_x - start_x
    segment_width = total_width / count

    for i, color in enumerate(final_colors):
        pen = QPen(color)
        pen.setWidthF(thickness)
        pen.setCapStyle(Qt.PenCapStyle.FlatCap)
        painter.setPen(pen)

        seg_start = start_x + (i * segment_width)
        seg_end = start_x + ((i + 1) * segment_width)
        line_start_x = seg_start + arc_radius if i == 0 else seg_start
        line_end_x = seg_end - arc_radius if i == count - 1 else seg_end

        if i == 0:
            left_rect = QRectF(
                start_x,
                base_y - 2 * arc_radius,
                2 * arc_radius,
                2 * arc_radius,
            )
            painter.drawArc(left_rect, 180 * 16, 90 * 16)

        if line_end_x > line_start_x:
            painter.drawLine(
                QPointF(line_start_x, base_y),
                QPointF(line_end_x, base_y),
            )

        if i == count - 1:
            right_rect = QRectF(
                end_x - 2 * arc_radius,
                base_y - 2 * arc_radius,
                2 * arc_radius,
                2 * arc_radius,
            )
            painter.drawArc(right_rect, 270 * 16, 90 * 16)

File: cpp/toolkit/src/helpers/underline_painter.cpp
#include "sli/toolkit/helpers/underline_painter.h"
#include "sli/toolkit/theme.h"

#include <QLineEdit>
#include <QPainter>
#include <QPen>
#include <QPointF>
#include <QWidget>

#include <algorithm>
#include <cmath>
#include <numbers>

namespace sli::toolkit {

namespace {

constexpr int kTaperSegments = 24;

double widgetScale(const QRectF& rect)
{
    double h = static_cast<double>(rect.height());
    return std::max(1.0, h / 32.0);
}

void drawTaperedArc(
    QPainter& painter,
    const QColor& baseColor,
    double thickness,
    double cx,
    double cy,
    double radius,
    double startDeg,
    double sweepDeg,
    int fullAlpha,
    double alphaAtStart,
    double alphaAtEnd)
{
    QPen pen(baseColor);
    pen.setWidthF(thickness);
    pen.setCapStyle(Qt::FlatCap);

    std::optional<QPointF> prevPt;

    for (int s = 0; s <= kTaperSegments; ++s) {
        double t = static_cast<double>(s) / kTaperSegments;
        double ang = std::numbers::pi / 180.0 * (startDeg + sweepDeg * t);
        double x = cx + radius * std::cos(ang);
        double y = cy - radius * std::sin(ang);
        QPointF pt(x, y);

        if (prevPt.has_value()) {
            double tMid = (s - 0.5) / kTaperSegments;
            double alphaNorm = alphaAtStart + (alphaAtEnd - alphaAtStart) * tMid;
            int alphaVal = static_cast<int>(
                std::round(fullAlpha * std::clamp(alphaNorm, 0.0, 1.0)));
            QColor segColor(baseColor);
            segColor.setAlpha(alphaVal);
            pen.setColor(segColor);
            painter.setPen(pen);
            painter.drawLine(prevPt.value(), pt);
        }
        prevPt = pt;
    }
}

} // anonymous namespace

void drawBottomUnderline(
    QPainter& painter,
    const QRectF& rect,
    const UnderlineConfig& config,
    QWidget* deviceWidget)
{
    // Python dark-mode guard: in dark mode only draw on QLineEdit widgets.
    if (Theme::isDark()) {
        if (!deviceWidget || !qobject_cast<QLineEdit*>(deviceWidget))
            return;
    }

    // Build color list.
    std::vector<QColor> colors;
    if (config.color.has_value()) {
        colors.push_back(config.color.value());
    } else if (!config.colors.empty()) {
        colors = config.colors;
    } else {
        colors.emplace_back(Theme::getColor("button.default.bottom.edge"));
    }

    // Apply alpha override.
    if (config.alpha.has_value()) {
        for (auto& c : colors) {
            c.setAlpha(config.alpha.value());
        }
    }

    int count = static_cast<int>(colors.size());
    if (count == 0)
        return;

    double scale = widgetScale(rect);
    double arcRadius = config.arcRadius * scale;
    double thickness = config.thickness;
    double verticalOffset = config.verticalOffset * scale;

    double baseY = rect.bottom() - verticalOffset;
    double startX = rect.left();
    double endX = rect.right();
    double totalWidth = endX - startX;
    double segmentWidth = totalWidth / count;

    for (int i = 0; i < count; ++i) {
        const QColor& color = colors[i];

        QPen pen(color);
        pen.setWidthF(thickness);
        pen.setCapStyle(Qt::FlatCap);
        painter.setPen(pen);

        double segStart = startX + i * segmentWidth;
        double segEnd = startX + (i + 1) * segmentWidth;
        double lineStartX = (i == 0) ? (segStart + arcRadius) : segStart;
        double lineEndX = (i == count - 1) ? (segEnd - arcRadius) : segEnd;

        if (lineEndX > lineStartX) {
            painter.drawLine(QPointF(lineStartX, baseY), QPointF(lineEndX, baseY));
        }

        int fullAlpha = color.alpha();
        if (arcRadius <= 0.0)
            continue;

        if (i == 0) {
            // Left end: arc 180°→270°, alpha fades 0→full
            double cxx = startX + arcRadius;
            double cyy = baseY - arcRadius;
            drawTaperedArc(painter, color, thickness, cxx, cyy, arcRadius,
                           180.0, 90.0, fullAlpha, 0.0, 1.0);
        }

        if (i == count - 1) {
            // Right end: arc 270°→360°, alpha fades full→0
            double cxx = endX - arcRadius;
            double cyy = baseY - arcRadius;
            drawTaperedArc(painter, color, thickness, cxx, cyy, arcRadius,
                           270.0, 90.0, fullAlpha, 1.0, 0.0);
        }
    }
}

}  // namespace sli::toolkit